### PR TITLE
PLFM-6298: Process expired approvals

### PR DESCRIPTION
**Pull request by marcomarasca**

### CODE REVIEW

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOAccessApprovalDAOImpl.java [hunk 1]
```diff
@@ -387,11 +406,62 @@ public static String buildAccessorGroupQuery(String accessRequirementId, String
...
+		ValidateArgument.required(expiredAfter, "expiredAfter");
+		ValidateArgument.requirement(limit > 0, "The limit must be greater than 0.");
+		
+		return jdbcTemplate.queryForList(SELECT_EXPIRED_APPROVALS, Long.class,
+				expiredAfter.toEpochMilli(), 
+				Instant.now().toEpochMilli(),
+				limit);
+	}
+	
+	@Override
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/dao/dataaccess/DBOAccessApprovalDAOImpl.java]: @transaction

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/DBOAccessApprovalDAOImplTest.java [hunk 2]
```diff
@@ -408,4 +380,458 @@ public void testBuildQuery() {
...
+		Long userId = Long.valueOf(individualGroup.getId());
+		List<Long> ids = Arrays.asList(accessApproval.getId(), accessApproval2.getId());
+		
+		List<Long> expected = ids;
+		
+		// Call under test
+		List<Long> result = accessApprovalDAO.revokeBatch(userId, ids);
+		
+		assertEquals(expected, result);
+
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/dao/DBOAccessApprovalDAOImplTest.java:560]: check etag changed.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessApprovalManagerImpl.java [hunk 3]
```diff
@@ -172,4 +212,63 @@ public BatchAccessApprovalInfoResponse getAccessApprovalInfo(UserInfo userInfo,
...
+	
+	@Override
+	@WriteTransaction
+	public int revokeExpiredApprovals(UserInfo userInfo, Instant expiredAfter, int maxBatchSize) {
+		ValidateArgument.required(userInfo, "The user");
+		ValidateArgument.required(expiredAfter, "The expiredAfter");
+		ValidateArgument.requirement(maxBatchSize > 0, "The maxBatchSize must be greater than 0.");
+		ValidateArgument.requirement(expiredAfter.isBefore(Instant.now()), "The expiredAfter must be a value in the past.");
+	
+		// Fetch the list of expired approval
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessApprovalManagerImpl.java:255]: check for admin

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessApprovalManagerImpl.java [hunk 4]
```diff
@@ -172,4 +212,63 @@ public BatchAccessApprovalInfoResponse getAccessApprovalInfo(UserInfo userInfo,
...
+		if (!authorizationManager.isACTTeamMemberOrAdmin(userInfo)) {
+			throw new UnauthorizedException("Only ACT member can perform this action.");
+		}
+		
+		if (accessorIds.isEmpty()) {
+			return;
+		}
+		
+		final List<Long> approvals = accessApprovalDAO.listApprovalsBySubmitter(accessRequirementId, submitterId, accessorIds);
+		final List<Long> revokedApprovals = accessApprovalDAO.revokeBatch(userInfo.getId(), approvals);
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/dataaccess/AccessApprovalManagerImpl.java:238]: empty check on approvals

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/dataaccess/AccessApprovalManagerImplUnitTest.java [hunk 5]
```diff
@@ -352,4 +457,268 @@ public void testGetAccessApprovalInfo() {
...
+		String accessRequirementId = "2";
+		String submitterId = "1";
+		List<String> accessorIds = Arrays.asList("1", "2");
+		
+		// Call under test
+		manager.revokeGroup(userInfo, accessRequirementId, submitterId, accessorIds);
+		
+		verify(mockAccessApprovalDAO).listApprovalsBySubmitter(accessRequirementId, submitterId, accessorIds);
+		verify(mockAccessApprovalDAO).revokeBatch(userInfo.getId(), approvals);
+
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/dataaccess/AccessApprovalManagerImplUnitTest.java:654]: verify isACT

File: services/workers/src/test/java/org/sagebionetworks/dataaccess/workers/AccessApprovalExpirationWorkerIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+		AccessApproval ap2 = newApproval(newAccessRequirement(), ApprovalState.REVOKED, Instant.now().minus(1, ChronoUnit.DAYS));
+		// APPROVED by expired
+		AccessApproval ap3 = newApproval(newAccessRequirement(), ApprovalState.APPROVED, Instant.now().minus(1, ChronoUnit.DAYS));
+		// Already revoked and not expiring
+		AccessApproval ap4 = newApproval(newAccessRequirement(), ApprovalState.REVOKED, null);
+		// Approved and expiring in the future
+		AccessApproval ap5 = newApproval(newAccessRequirement(), ApprovalState.APPROVED, Instant.now().plus(1, ChronoUnit.DAYS));
+		
+		// Simulates the run
+		worker.run(null);
```
Comment (john-hill) [inline services/workers/src/test/java/org/sagebionetworks/dataaccess/workers/AccessApprovalExpirationWorkerIntegrationTest.java]: will move to manager integration test, and for the worker just a simple did something get revoked.

File: services/workers/src/main/resources/dataaccess-worker-spb.xml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+						</constructor-arg>
+					</bean>
+				</property>
+				<property name="targetMethod" value="run" />
+				<property name="concurrent" value="false" />
+			</bean>
+		</property>
+		<property name="startDelay" value="1012" />
+		<!-- Once every 10 minutes -->
+		<property name="repeatInterval" value="600000" />
```
Comment (john-hill) [inline services/workers/src/main/resources/dataaccess-worker-spb.xml]: can be 30 secs without loop

**Comment by marcomarasca** [comment]
I disabled in the main-scheduler-spb.xml the trigger for the worker that automatically revokes the access approvals for now, so this should be mergeable. In the next PR I'll add the actual sending of notifications so that they can go together.

**Review by john-hill — APPROVED** [review]

**Pull request merged**